cpp/include/cudf/io/detail/data_casting.cuh

```diff
@@ -366,20 +841,102 @@ std::unique_ptr<column> parse_data(str_tuple_it str_tuples,
   auto null_count_data = d_null_count.data();
 
   if (col_type == cudf::data_type{cudf::type_id::STRING}) {
-    // this utility calls the functor to build the offsets and chars columns;
-    // the bitmask and null count may be updated by parse failures
-    auto [offsets, chars] = cudf::strings::detail::make_strings_children(
-      string_parse<decltype(str_tuples)>{
-        str_tuples, static_cast<bitmask_type*>(null_mask.data()), null_count_data, options},
-      col_size,
-      stream,
-      mr);
-
-    return make_strings_column(col_size,
-                               std::move(offsets),
-                               std::move(chars),
-                               d_null_count.value(stream),
-                               std::move(null_mask));
+    auto const max_length = thrust::transform_reduce(
+      rmm::exec_policy(stream),
+      str_tuples,
+      str_tuples + col_size,
+      [] __device__(auto t) { return t.second; },
+      size_type{0},
+      thrust::maximum<size_type>{});
+
+    if (max_length < SINGLE_THREAD_THRESHOLD) {
+      // this utility calls the functor to build the offsets and chars columns;
+      // the bitmask and null count may be updated by parse failures
+      nvtxRangePush("make_strings_children");
+      auto [offsets, chars] = cudf::strings::detail::make_strings_children(
+        string_parse<decltype(str_tuples)>{
+          str_tuples, static_cast<bitmask_type*>(null_mask.data()), null_count_data, options},
+        col_size,
+        stream,
+        mr);
+      nvtxRangePop();
+
+      return make_strings_column(col_size,
+                                 std::move(offsets),
+                                 std::move(chars),
+                                 d_null_count.value(stream),
+                                 std::move(null_mask));
+    } else {
+      nvtxRangePush("string_parallel");
+      auto offsets2 = cudf::make_numeric_column(
+        data_type{cudf::type_id::INT32}, col_size + 1, cudf::mask_state::UNALLOCATED, stream, mr);
+      auto d_offsets = offsets2->mutable_view().data<size_type>();
+
+      constexpr auto warps_per_block  = 8;
+      constexpr int threads_per_block = cudf::detail::warp_size * warps_per_block;
+      auto num_blocks                 = min(65535, col_size / warps_per_block + 1);
+      auto str_counter                = cudf::numeric_scalar(size_type{0}, true, stream);
+
+      parse_fn_warp_parallel<<<num_blocks, threads_per_block, 0, stream.value()>>>(
```

Comment:
`parse_fn_warp_parallel` and `parse_fn_block_parallel` look independent. Might improve performance to run them in different stream. I think having a static stream pool with two streams would be sufficient here to help overlap some of the work.

Reply:
do we already use static stream pool in io code?

Reply:
not yet. I hope it will be merged soon, in https://github.com/rapidsai/cudf/pull/13922.
We can leave a TODO until the global pool lands.

Reply:
Left a TODO.
stream pool did not achieve signficant improvement. So, will experiment more for another PR. Not including it in this PR.
`./_deps/nvbench-src/scripts/nvbench_compare.py ./c2.json c4.json `
['./c2.json', 'c4.json']
# json_read_string_column

## [0] Quadro GV100

|  max_length  |  is_fixed_length  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |         Diff |   %Diff |  Status  |
|--------------|-------------------|------------|-------------|------------|-------------|--------------|---------|----------|
|      16      |         1         | 244.442 ms |       0.21% | 245.402 ms |       0.96% |   959.972 us |   0.39% |   FAIL   |
|      32      |         1         | 160.267 ms |       0.18% | 160.262 ms |       0.09% |    -5.881 us |  -0.00% |   PASS   |
|      64      |         1         | 161.867 ms |       0.44% | 161.403 ms |       0.06% |  -463.632 us |  -0.29% |   FAIL   |
|     128      |         1         | 162.441 ms |       0.36% | 161.767 ms |       0.10% |  -673.526 us |  -0.41% |   FAIL   |
|     256      |         1         | 144.881 ms |       0.47% | 144.819 ms |       0.26% |   -62.016 us |  -0.04% |   PASS   |
|     512      |         1         | 132.372 ms |       0.48% | 131.672 ms |       0.07% |  -700.235 us |  -0.53% |   FAIL   |
|     1024     |         1         | 125.344 ms |       0.37% | 125.133 ms |       0.05% |  -211.272 us |  -0.17% |   FAIL   |
|     4096     |         1         | 121.457 ms |       0.50% | 120.974 ms |       0.08% |  -483.184 us |  -0.40% |   FAIL   |
|    16384     |         1         | 135.223 ms |       0.49% | 134.880 ms |       0.47% |  -343.044 us |  -0.25% |   PASS   |
|    65536     |         1         | 142.435 ms |       0.27% | 142.837 ms |       0.49% |   402.222 us |   0.28% |   FAIL   |
|    262144    |         1         | 150.038 ms |       0.41% | 150.162 ms |       0.43% |   123.804 us |   0.08% |   PASS   |
|   1048576    |         1         | 347.749 ms |       0.29% | 347.054 ms |       0.12% |  -695.074 us |  -0.20% |   FAIL   |
|   10485760   |         1         |    2.737 s |       0.08% |    2.735 s |       0.13% | -1876.563 us |  -0.07% |   PASS   |
|  104857600   |         1         |   18.052 s |        inf% |   18.056 s |        inf% |     4.113 ms |   0.02% |   PASS   |
|      16      |         0         | 338.887 ms |       0.18% | 337.814 ms |       0.07% | -1073.114 us |  -0.32% |   FAIL   |
|      32      |         0         | 202.162 ms |       0.26% | 203.290 ms |       0.48% |     1.128 ms |   0.56% |   FAIL   |
|      64      |         0         | 148.158 ms |       0.46% | 148.238 ms |       0.45% |    80.219 us |   0.05% |   PASS   |
|     128      |         0         | 188.805 ms |       0.14% | 188.934 ms |       0.14% |   129.172 us |   0.07% |   PASS   |
|     256      |         0         | 153.771 ms |       0.50% | 153.039 ms |       0.04% |  -731.565 us |  -0.48% |   FAIL   |
|     512      |         0         | 141.374 ms |       0.10% | 141.641 ms |       0.29% |   266.837 us |   0.19% |   FAIL   |
|     1024     |         0         | 132.822 ms |       0.49% | 132.785 ms |       0.48% |   -36.966 us |  -0.03% |   PASS   |
|     4096     |         0         | 126.245 ms |       0.47% | 126.147 ms |       0.18% |   -98.391 us |  -0.08% |   PASS   |
|    16384     |         0         | 128.630 ms |       0.30% | 126.918 ms |       0.49% | -1711.989 us |  -1.33% |   FAIL   |
|    65536     |         0         | 145.012 ms |       0.17% | 146.203 ms |       0.37% |     1.191 ms |   0.82% |   FAIL   |
|    262144    |         0         | 145.651 ms |       0.10% | 145.060 ms |       0.50% |  -590.480 us |  -0.41% |   FAIL   |
|   1048576    |         0         | 292.832 ms |       0.11% | 293.013 ms |       0.34% |   180.487 us |   0.06% |   PASS   |
|   10485760   |         0         |    2.323 s |       0.02% |    2.323 s |       0.05% |    -8.594 us |  -0.00% |   PASS   |
|  104857600   |         0         |   10.164 s |        inf% |   10.164 s |        inf% |   130.859 us |   0.00% |   PASS   |

Reply:
Darn. It would be good to take a look at a profile with/without streams.